File: dotnet/Quill/DriverCreation.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;
using OpenQA.Selenium.Chromium;
using OpenQA.Selenium.Edge;
using OpenQA.Selenium.Firefox;
using OpenQA.Selenium.IE;
using OpenQA.Selenium.Safari;

namespace Quill
{
    public static class DriverCreation
    {
        public static BrowserType type = BrowserType.Firefox;

        private static List<string> flags = new List<string>();

        public static DriverOptions options = new DriverOptions();

        public static string ExecutablePath { get; private set; }

        public static LogLevel logLevel { get; set; }
        public static bool SuppressInitialDiagnosticInformation { get; set; }

        public static void AddFlag(string flag)
        {
            flags.Add(flag);
        }

        public static void SetExecutablePath(string path)
        {
            ExecutablePath = path;
        }

        public static void SetBrowserType(BrowserType type)
        {
            DriverCreation.type = type;
        }

        public static WebDriver CreateNew()
        {
            return CreateNew(type);
        }

        public static WebDriver CreateNew(BrowserType type)
        {
            switch (type)
            {
                case BrowserType.Chrome:
                    {                        
                        ChromeOptions cOptions = new ChromeOptions();
                        ChromeDriverService cDriverService;

                        if (ExecutablePath == string.Empty)
                            cDriverService = ChromeDriverService.CreateDefaultService();
                        else
                            cDriverService = ChromeDriverService.CreateDefaultService(ExecutablePath);

                        if (options.headless)
                        {
                            cOptions.AddArgument("--headless=new");
                            cOptions.AddArgument("--window-position=-32000,-32000");
                            cOptions.AddArgument("--log-level=3");

                        }
                        if (!options.loadImages)
                            cOptions.AddArgument("blink-settings=imagesEnabled=false");
                        
                        cOptions.AddArgument("--window-size=2560,1440");                        

                        foreach (string flag in flags)
                        {
                            cOptions.AddArgument(flag);
                        }

                        cDriverService.HideCommandPromptWindow = true;
                        cDriverService.SuppressInitialDiagnosticInformation = SuppressInitialDiagnosticInformation;
                        return new ChromeDriver(cDriverService, cOptions);
                    }
                case BrowserType.InternetExplorer:
                    {
                        InternetExplorerOptions ieOptions = new InternetExplorerOptions();
                        var ieDriverService = InternetExplorerDriverService.CreateDefaultService();

                        if (options.headless)
                        {
                            ieOptions.IntroduceInstabilityByIgnoringProtectedModeSettings = true;
                            ieOptions.EnsureCleanSession = true;
                            ieOptions.EnableNativeEvents = false;
                            ieOptions.BrowserCommandLineArguments = "-private"; // Start IE in private mode
                            ieOptions.IntroduceInstabilityByIgnoringProtectedModeSettings = true;
                        }

                        foreach (string flag in flags)
                        {
                            ieOptions.AddAdditionalOption(flag, true);
                        }

                        ieDriverService.HideCommandPromptWindow = true;


                        return new InternetExplorerDriver(ieDriverService, ieOptions);
                    }
                    
                case BrowserType.Edge:
                    {
                        EdgeOptions eOptions = new EdgeOptions();
                        var eDriverService = EdgeDriverService.CreateDefaultService();

                        if (options.headless)
                            eOptions.AddArgument("headless");

                        if (!options.loadImages)
                            eOptions.AddArgument("blink-settings=imagesEnabled=false");

                        foreach (string flag in flags)
                        {
                            eOptions.AddArgument(flag);
                        }

                        eDriverService.HideCommandPromptWindow = true;

                        return new EdgeDriver(eDriverService, eOptions);
                    }
                case BrowserType.Safari:
                    {
                        SafariOptions sOptions = new SafariOptions();
                        var sDriverService = SafariDriverService.CreateDefaultService();

                        foreach (string flag in flags)
                        {
                            sOptions.AddAdditionalOption(flag, true);
                        }
                        sDriverService.HideCommandPromptWindow = true;

                        return new SafariDriver(sDriverService, sOptions);
                    }
                case BrowserType.Firefox:
                    {
                        FirefoxOptions fOptions = new FirefoxOptions();
                        var fDriverService = FirefoxDriverService.CreateDefaultService();

                        if (ExecutablePath != string.Empty)
                            fDriverService = FirefoxDriverService.CreateDefaultService(ExecutablePath);                                                

                        if (options.headless)
                            fOptions.AddArgument("--headless");
                        if (!options.loadImages)
                            fOptions.AddArgument("blink-settings=imagesEnabled=false");

                        foreach (string flag in flags)
                        {
                            fOptions.AddArgument(flag);
                        }

                        if(logLevel != LogLevel.None)
                            fOptions.LogLevel = (FirefoxDriverLogLevel)logLevel;
                        else
                        {
                            Environment.SetEnvironmentVariable("MOZ_LOG", ""); 
                            Environment.SetEnvironmentVariable("MOZ_LOG_FILE", "/dev/null");
                        }
                        //fOptions.LogLevel = FirefoxDriverLogLevel.Fatal;

                        fDriverService.HideCommandPromptWindow = true;
                        fDriverService.SuppressInitialDiagnosticInformation = SuppressInitialDiagnosticInformation;
                        return new FirefoxDriver(fDriverService, fOptions);
                    }
            }

            return new ChromeDriver();
        }

    }

    public struct DriverOptions
    {
        /// <summary>
        /// Prevent the browser from appearing
        /// </summary>
        /// <remarks>
        /// True by default
        /// </remarks>
        public bool headless;

        /// <summary>
        /// lets the browser load images.  (May break media uploading)
        /// </summary>
        /// <remarks>
        /// True by default
        /// </remarks>
        public bool loadImages;

        public DriverOptions()
        {
            headless = true;
            loadImages = true;
        }
    }

    public enum BrowserType
    {
        /// <summary>
        /// Chrome does not support emojis as of right now
        /// </summary>
        Chrome,
        /// <summary>
        /// Not recommended. Requires your operating system to be in English. Driver options may not function properly
        /// </summary>
        InternetExplorer,
        Edge,
        /// <summary>
        /// Not tested. All flags will be ignored
        /// </summary>
        Safari,

        /// <summary>
        /// The best option for most use cases
        /// </summary>
        Firefox
    }

    public enum LogLevel
    {
        /// <summary>
        /// Represents the Trace value, the most detailed logging level available.
        /// </summary>        
        Trace,
        
        /// <summary>
        /// Represents the Debug value
        /// </summary>
        Debug,        

        /// <summary>
        /// Represents the Config value
        /// </summary>
        Config,
        
        /// <summary>
        /// Represents the Info value
        /// </summary>
        Info,
        
        /// <summary>
        /// Represents the Warn value
        /// </summary>
        Warn,
        
        /// <summary>
        /// Represents the Error value
        /// </summary>
        Error,
        
        /// <summary>         
        /// Represents the Fatal value, the least detailed logging level available.
        /// </summary>
        Fatal,        
          
        /// <summary>
        /// Represents that the logging value is unspecified, and should be the default level.
        /// </summary>
        Default,

        /// <summary>
        /// Makes the driver's output pipe to a null address
        /// </summary>
        None,
    }
}

File: javascript/Quill/DriverCreation.ts
import chrome from 'selenium-webdriver/chrome';
import firefox from 'selenium-webdriver/firefox';
import ie from 'selenium-webdriver/ie';
import { Builder, WebDriver } from 'selenium-webdriver';
import { List } from './tools';

enum BrowserType 
{
    /** **Default** Chrome curently Dose Not Support Emojis
     */
    Chrome,
    /** **Not Recommended.** Requires your operating system to be in English. Driver options may not function properly
     */
    InternetExplorer,
    Edge,
    /** **Not Recomended** All flages will be ignored
     */
    Safari,
    Firefox,
}

/**
 * creates this awsome sauce driver
 */
export class DriverOptions {
    /**
     * Prevent the browser from appearing
     * @remarks True by default
     */
    public headless: boolean;

    /**
     * lets the browser load images. (May break media uploading)
     * @remarks True by default
     */
    public loadImages: boolean;

    /**
     * Prevent the browser from loading css
     * @remarks False by default
     */
    public disableCSS: boolean;

    constructor() 
    {
        this.headless = true;
        this.loadImages = true;
        this.disableCSS = false;
    }
}

export default class DriverCreation {
    public static type: BrowserType = BrowserType.Chrome;

    public static options: DriverOptions = new DriverOptions();

    public static executablePath: string;

    private static flags: List<string> = new List<string>();

    public static AddFlag(flag: string): void 
    {
        this.flags.add(flag);
    }

    public static SetExecutablePath(path: string): void 
    {
        this.executablePath = path;
    }

    public static SetBrowserType(type: BrowserType): void 
    {
        this.type = type;
    }

    public static CreateNew(): WebDriver; // << SIG
    public static CreateNew(type: BrowserType): WebDriver; // << SIG
    public static CreateNew(type?: BrowserType): WebDriver  // << IMPL
    {
        switch (type)
        {
            case BrowserType.Chrome:
                {
                    var cOptions = new chrome.Options();
                    var cDriverService: chrome.ServiceBuilder;

                    if (!this.executablePath)
                        cDriverService = new chrome.ServiceBuilder();
                    else
                        cDriverService = new chrome.ServiceBuilder(this.executablePath);

                    if (this.options.headless)
                        cOptions.addArguments("--headless=new", "--window-position=-32000,-32000", "--log-level=3",);
                    if (!this.options.loadImages)
                        cOptions.addArguments("--blink-settings=imagesEnabled=false");
                    cOptions.addArguments("--window-size=2560,1440");

                    for (const flag of this.flags)
                    {
                        cOptions.addArguments(flag);
                    }

                    return new Builder()
                        .forBrowser('chrome')
                        .setChromeService(cDriverService)
                        .setChromeOptions(cOptions)
                        .build();
                }
        case BrowserType.Firefox:
            {
                var fOptions = new firefox.Options();
                var fDriverService: firefox.ServiceBuilder;

                if (!this.executablePath)
                    fDriverService = new firefox.ServiceBuilder();
                else
                    fDriverService = new firefox.ServiceBuilder(this.executablePath);

                if (this.options.headless)
                    fOptions.addArguments("--headless");
                if (!this.options.loadImages)
                    fOptions.addArguments("--disable-images");

                for (const flag of this.flags) {
                    fOptions.addArguments(flag);
                }

                return new Builder()
                    .forBrowser('firefox')
                    .setFirefoxService(fDriverService)
                    .setFirefoxOptions(fOptions)
                    .build();
            }
        case BrowserType.Edge:
            {
                var eOptions = new chrome.Options();
                var eDriverService: chrome.ServiceBuilder;

                if (!this.executablePath)
                    eDriverService = new chrome.ServiceBuilder();
                else
                    eDriverService = new chrome.ServiceBuilder(this.executablePath);

                if (this.options.headless)
                    eOptions.addArguments("--headless=new");
                if (!this.options.loadImages)
                    eOptions.addArguments("--blink-settings=imagesEnabled=false");

                for (const flag of this.flags) {
                    eOptions.addArguments(flag);
                }

                return new Builder()
                    .forBrowser('MicrosoftEdge')
                    .setChromeService(eDriverService)
                    .setChromeOptions(eOptions)
                    .build();
            }
        case BrowserType.InternetExplorer:
            {
                var ieOptions = new ie.Options();
                var ieDriverService: ie.ServiceBuilder;


                if (!this.executablePath)
                    ieDriverService = new ie.ServiceBuilder();
                else
                    ieDriverService = new ie.ServiceBuilder(this.executablePath);

                if (this.options.headless)
                {
                    ieOptions.introduceFlakinessByIgnoringProtectedModeSettings(true);
                    ieOptions.ensureCleanSession(true);
                    ieOptions.addArguments("--private");
                    ieOptions.introduceFlakinessByIgnoringProtectedModeSettings(true);
                }

                for (const flag of this.flags) {
                    ieOptions.addArguments(flag);
                }

                return new Builder()
                    .forBrowser('ie')
                    .setIeService(ieDriverService)
                    .setIeOptions(ieOptions)
                    .build();
            }
        case BrowserType.Safari:
            throw new Error(`Browser type ${BrowserType[type]} is not supported`);
        default:
            return this.CreateNew(BrowserType.Chrome);

        }

    }
}




